stdlib/buffer.ml

```diff
@@ -87,102 +96,91 @@ let resize b more =
   let new_buffer = Bytes.create !new_len in
   (* PR#6148: let's keep using [blit] rather than [unsafe_blit] in
      this tricky function that is slow anyway. *)
-  Bytes.blit b.buffer 0 new_buffer 0 b.position;
-  b.buffer <- new_buffer;
-  b.length <- !new_len;
-  assert (b.position + more <= b.length);
-  assert (old_pos + more <= b.length);
+  Bytes.blit b.inner.buffer 0 new_buffer 0 b.position;
```

Comment:
I think that we could save the whole inner buffer at the start of this function, and use it here as the source instead of reloading. The length (`old_len` here) would also be used instead of `b.position` to specify the number of bytes to copy. This would avoid potential exceptions from `blit` when concurrent modification occurs, at the cost of being more likely to drop concurrently added data. We could even switch back to `unsafe_blit` in this case, although I would argue that it's better to be safe in this non-critical part of the code.

Reply:
This sounds like a good idea in term of code regularity (and in term of regular behavior in presence of a data race).

Do you think it is really worth it to use `old_len` rather than `old_pos` with the aim of minimizing dropping unsynchronized data?

Any opinion on this "specification" change @gasche?

Reply:
I was also tempted to suggest to reuse `old_pos` instead of `b.position`, but I decided to not make the point explicitly -- you only get a *fraction* of my nitpick comments!

For me the problem with the assert is not that they fail (I think it's actually better to fail if we detect this kind of races), but that the code reads as if we believed that they will never fail. (You removed a lengthy comment that is mostly useless now, but it contained this useful information about the assertion that we know they may fail.)

It's actually fun to wonder whether we could detect all resize/resize races, fail with an error in this case, and behave somewhat-nicely -- at no performance cost. I suppose that suggestions similar to the "generation count" from the dynarray discussion could do it. But this could be left to a follow-up PR.

Reply:
If you stick to `b.inner.buffer` in the call to `blit` then you're already in a case where you don't know the length of the buffer, so re-fetching `b.position` is the solution least likely to drop concurrently added data (and most likely to be in sync with the actual buffer contents in general).
If you want to avoid out-of-bound accesses as much as possible, then blitting from the old buffer and `old_pos` is safer, even safer with `old_len` instead of `old_pos` (you guard against a reset occurring between fetching `b.position` and `b.inner`), but you're guaranteed to drop any concurrent updates occurring while you're computing the size and allocating the new array.

Reply:
Then let's stick to just removing the assertions for this PR.